src/log-store/src/kafka/producer.rs

```diff
@@ -0,0 +1,470 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use common_telemetry::{debug, warn};
+use futures::future::try_join_all;
+use rskafka::client::partition::Compression;
+use rskafka::client::producer::ProducerClient;
+use rskafka::record::Record;
+use snafu::ResultExt;
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::sync::oneshot;
+
+use crate::error::{self, Result};
+
+pub struct ProduceRequest {
+    batch: Vec<Record>,
+    sender: oneshot::Sender<ProduceResultReceiver>,
+}
+
+#[derive(Default)]
+struct ProduceResultReceiver {
+    receivers: Vec<oneshot::Receiver<Result<Vec<i64>>>>,
+}
+
+impl ProduceResultReceiver {
+    fn add_receiver(&mut self, receiver: oneshot::Receiver<Result<Vec<i64>>>) {
+        self.receivers.push(receiver)
+    }
+
+    async fn wait(self) -> Result<Vec<i64>> {
+        Ok(try_join_all(self.receivers)
+            .await
+            .into_iter()
+            .flatten()
+            .collect::<Result<Vec<_>>>()?
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>())
+    }
+}
+
+struct BackgroundProducerWorker {
+    /// The [`ProducerClient`].
+    client: Arc<dyn ProducerClient>,
+    // The compression configuration.
+    compression: Compression,
+    // The running flag.
+    running: Arc<AtomicBool>,
+    /// Receiver of [ProduceRequest].
+    receiver: Receiver<ProduceRequest>,
+    /// Max batch size for a worker to handle requests.
+    request_batch_size: usize,
+    /// Max bytes size for a single flush.
+    max_batch_bytes: usize,
+    /// The [PendingRequest]s.
+    pending_requests: Vec<PendingRequest>,
+}
+
+struct PendingRequest {
+    batch: Vec<Record>,
+    size: usize,
+    sender: oneshot::Sender<Result<Vec<i64>>>,
+}
+
+/// ## Panic
+/// Panic if any [Record]'s `approximate_size` > `max_batch_bytes`.
+fn handle_produce_requests(
+    requests: &mut Vec<ProduceRequest>,
+    max_batch_bytes: usize,
+) -> Vec<PendingRequest> {
+    let mut records_buffer = vec![];
+    let mut batch_size = 0;
+    let mut pending_requests = Vec::with_capacity(requests.len());
+
+    for ProduceRequest { batch, sender } in requests.drain(..) {
+        let mut receiver = ProduceResultReceiver::default();
+        for record in batch {
+            assert!(record.approximate_size() <= max_batch_bytes);
```

Comment:
We need to ensure that the metadata size < [ESTIMATED_META_SIZE](https://github.com/fengjiachun/greptimedb/blob/637a271e70e1f686fa5c9f564a10ead765d1fd81/src/log-store/src/kafka/util/record.rs#L36)